A77-xds-rate-limiting-rlqs.md

```diff
@@ -0,0 +1,618 @@
+corresponding timer is removed.
+
+### Integrations
+
+#### Connecting to RLQS Control Plane
+
+xDS Control Plane provides RLQS connection details in [`GrpcService`] message (
+already supported by Envoy). `GrpcService` supports two modes:
+
+1. `GrpcService.EnvoyGrpc`, Envoy's minimal custom gRPC client implementation.
+2. [`GrpcService.GoogleGrpc`], regular gRPC-cpp client.
+
+For obvious reasons, we'll only support the `GoogleGrpc` mode.
+
+##### Security Considerations
+
+In [`GrpcService.GoogleGrpc`], the xDS Control Plane provides the target URI,
+`channel_credentials`, and `call_credentials`. If the xDS Control Plane is
+compromised, an attacker could configure the xDS clients to talk to other
+malicious Control Planes. This could lead to potential exploits such as leaking
+customer's Application Default Credentials OAuth token.
+
+To prevent that, we'll introduce an allow-list to the bootstrap file introduced
+in [gRFC A27]. This allow-list will be a map from a fully-qualified server
+target URI to an object containing channel credentials to use for that target.
+
+```json5
+// The allowlist of Control Planes allowed to be configured via xDS.
+"allowed_grpc_services": {
+  // The key is fully-qualified server URI.
+  "dns:///xds.example.org:443": {
+    // The value is an object containing "channel_creds".
+    "channel_creds": [
+      // The format is identical to xds_servers.channel_creds.
+      {
+        "type": "string containing channel cred type",
+        "config": "optional JSON object containing config for the type"
+      }
+    ]
+  }
+}
+```
+
+When xDS Control Plane configures connection to another control plane
+via [`GrpcService.GoogleGrpc`] message, we'll inspect the allow-list for the
+matching target URI.
+
+1. If target URI is not present, we don't create the connection to the requested
+   Control Plane, and NACK the xDS resource.
+2. If target URI is present, we create the connection to the requested Control
+   Plane using the channel credentials provided in the bootstrap file. Transport
+   security configuration provided by the TD is ignored.
+
+> [!IMPORTANT]
+> This solution is not specific to RLQS, and should be used with any
+> other Control Planes configured via [`GrpcService`] message.
+
+#### Unified Matcher API
+
+RPCs will be matched into buckets using [Unified Matcher API] — an adaptable
+framework that can be used in any xDS component that needs matching features.
+
+Envoy provides two syntactically equivalent Unified Matcher
+definitions: [`envoy.config.common.matcher.v3.Matcher`](https://github.com/envoyproxy/envoy/blob/e3da7ebb16ad01c2ac7662758a75dba5cdc024ce/api/envoy/config/common/matcher/v3/matcher.proto)
+and [`xds.type.matcher.v3.Matcher`](https://github.com/cncf/xds/blob/b4127c9b8d78b77423fd25169f05b7476b6ea932/xds/type/matcher/v3/matcher.proto).
+We will only support the latter, which is the preferred version for all new APIs
```

Comment:
Any reason not to support both?  I think there are other places where we'll want to use the same matcher implementation in the future that use the old type, and it seems like we should be able to handle either one.

Reply:
I don't want to introduce a legacy dependency we have to support forever without a real need for it. 

Note that CEL libraries will keep supporting the legacy format, but they're being migrated to the canonical protos, f.e. https://github.com/google/cel-cpp/commit/aff86967fc811ed4605f099ee2f502d331d5bbe7.  This means canonical protos will be used internally, and legacy CEL operations will incur a performance overhead. 

As for java, the initial release of CEL-java only supported canonical CEL in the first place, however now it's possible to convert the legacy protos  to the canonical one, but this leads to adding separate dependencies (CEL Java V1alpha1 Utility, CEL Java Protobuf Adapter).

AFAIK Envoy will be upgrading its CEL intermediate lib to the canonical protos too.

Reply:
Also - we can always add the support for CEL if necessary. But we won't be able to remove it once the genie is out of the bottle.

Reply:
I've misunderstood the question. This was about supporting `envoy.config.common.matcher.v3.Matcher` and had nothing to do with CEL. 

Yes, we can support `envoy.config.common.matcher.v3.Matcher`. Would it make sense to note that we should unpack it as `xds.type.matcher.v3.Matcher`?

Reply:
Covered in https://github.com/sergiitk/grfc-proposal/blob/a77-rlqs/A77-xds-rate-limiting-rlqs.md#unified-matcher-api-support